cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3431 @@
+
+  int64_t partsupp_batches_generated() const {
+    return partsupp_batches_generated_.load();
+  }
+
+  Result<std::shared_ptr<Schema>> SetPartOutputColumns(
+      const std::vector<std::string>& cols) {
+    return SetOutputColumns(cols, kPartTypes, kPartNameMap, part_cols_);
+  }
+
+  Result<std::shared_ptr<Schema>> SetPartSuppOutputColumns(
+      const std::vector<std::string>& cols) {
+    return SetOutputColumns(cols, kPartsuppTypes, kPartsuppNameMap, partsupp_cols_);
+  }
+
+  Result<util::optional<ExecBatch>> NextPartBatch() {
+    size_t thread_index = thread_indexer_();
+    ThreadLocalData& tld = thread_local_data_[thread_index];
+    {
+      std::lock_guard<std::mutex> lock(part_output_queue_mutex_);
+      if (!part_output_queue_.empty()) {
+        ExecBatch batch = std::move(part_output_queue_.front());
+        part_output_queue_.pop();
+        return std::move(batch);
+      } else if (part_rows_generated_ == part_rows_to_generate_) {
+        return util::nullopt;
+      } else {
+        tld.partkey_start = part_rows_generated_;
+        tld.part_to_generate =
+            std::min(batch_size_, part_rows_to_generate_ - part_rows_generated_);
+        part_rows_generated_ += tld.part_to_generate;
+
+        int64_t num_ps_batches = PartsuppBatchesToGenerate(thread_index);
+        part_batches_generated_.fetch_add(1);
+        partsupp_batches_generated_.fetch_add(num_ps_batches);
+        ARROW_DCHECK(part_rows_generated_ <= part_rows_to_generate_);
+      }
+    }
+    tld.part.resize(PART::kNumCols);
+    std::fill(tld.part.begin(), tld.part.end(), Datum());
+    RETURN_NOT_OK(InitPartsupp(thread_index));
+
+    for (int col : part_cols_) RETURN_NOT_OK(kPartGenerators[col](thread_index));
+    for (int col : partsupp_cols_) RETURN_NOT_OK(kPartsuppGenerators[col](thread_index));
+
+    std::vector<Datum> part_result(part_cols_.size());
+    for (size_t i = 0; i < part_cols_.size(); i++) {
+      int col_idx = part_cols_[i];
+      part_result[i] = tld.part[col_idx];
+    }
+    if (!partsupp_cols_.empty()) {
+      std::vector<ExecBatch> partsupp_results;
+      for (size_t ibatch = 0; ibatch < tld.partsupp.size(); ibatch++) {
+        std::vector<Datum> partsupp_result(partsupp_cols_.size());
+        for (size_t icol = 0; icol < partsupp_cols_.size(); icol++) {
+          int col_idx = partsupp_cols_[icol];
+          partsupp_result[icol] = tld.partsupp[ibatch][col_idx];
+        }
+        ARROW_ASSIGN_OR_RAISE(ExecBatch eb, ExecBatch::Make(std::move(partsupp_result)));
+        partsupp_results.emplace_back(std::move(eb));
+      }
+      {
+        std::lock_guard<std::mutex> guard(partsupp_output_queue_mutex_);
+        for (ExecBatch& eb : partsupp_results) {
+          partsupp_output_queue_.emplace(std::move(eb));
+        }
+      }
+    }
+    return ExecBatch::Make(std::move(part_result));
+  }
+
+  Result<util::optional<ExecBatch>> NextPartSuppBatch() {
```

Comment:
I may be mistaken (this code is difficult to follow and you didn't provide any comments), but it looks like this is doing home-grown thread synchronization. Is there a strong reason not to use the existing `ThreadPool` facility?

Reply:
i will add comments. there is thread synchronization but i don't see how ThreadPool can help me avoid grabbing mutexes on shared objects.